src/disjoint_mut.rs

```diff
@@ -407,143 +430,139 @@ mod debug {
             }
-            let mut ranges = self.ranges.mutable.lock().unwrap();
-            for r in ranges.iter() {
-                if range.overlaps(r) {
-                    let thread = thread::current().id();
-                    panic!("{range:?} on thread {thread:?} overlaps with an existing mutable range: {r:#?}");
-                }
+            let mut mut_bounds = self.bounds.mutable.lock().unwrap();
+            for b in mut_bounds.iter() {
+                check_overlaps(&bounds, true, b, true);
             }
-            ranges.push(DisjointMutRange::new(range.to_bounds()));
+            mut_bounds.push(DisjointMutBounds::new(bounds));
         }
 
-        fn add_immut_range<R: BoundsExt>(&self, range: &R) {
-            let ranges = self.ranges.mutable.lock().unwrap();
-            for r in ranges.iter() {
-                if range.overlaps(r) {
-                    let thread = thread::current().id();
-                    panic!("{range:?} on thread {thread:?} overlaps with an existing mutable range: {r:#?}");
-                }
+        #[track_caller]
+        fn add_immut_bounds(&self, bounds: Bounds) {
+            let mut_bounds = self.bounds.mutable.lock().unwrap();
+            for b in mut_bounds.iter() {
+                check_overlaps(&bounds, false, b, true);
             }
-            self.ranges
+            self.bounds
                 .immutable
                 .lock()
                 .unwrap()
-                .push(DisjointMutRange::new(range.to_bounds()));
+                .push(DisjointMutBounds::new(bounds));
         }
 
-        fn remove_range(&self, range: (Bound<usize>, Bound<usize>), mutable: bool) {
-            let ranges = if mutable {
-                self.ranges.mutable.lock()
+        fn remove_bound(&self, bounds: &Bounds, mutable: bool) {
+            let all_bounds = if mutable {
+                self.bounds.mutable.lock()
             } else {
-                self.ranges.immutable.lock()
+                self.bounds.immutable.lock()
             };
-            let Ok(mut ranges) = ranges else {
+            let Ok(mut all_bounds) = all_bounds else {
                 // Another thread has panicked holding a range lock. We can't
                 // remove anything.
                 return;
             };
-            let idx = ranges
+            let idx = all_bounds
                 .iter()
-                .position(|r| r == &range)
+                .position(|r| r.bounds == *bounds)
                 .expect("Expected range {range:?} to be in the active ranges");
-            ranges.remove(idx);
+            all_bounds.remove(idx);
         }
     }
 
     impl<'a, T: AsMutPtr, V: ?Sized> DisjointMutGuard<'a, T, V> {
-        pub fn new(
-            parent: &'a DisjointMut<T>,
-            slice: &'a mut V,
-            range: (Bound<usize>, Bound<usize>),
-        ) -> Self {
-            parent.add_mut_range(&range);
+        #[track_caller]
```

Comment:
cfg_attr

Reply:
I'll leave these ones since it's in `mod debug` already.